ksim/losses/ppo.py

```diff
@@ -0,0 +1,175 @@
+            action_history=None,
+            recurrent_state=None,
+        )
+
+        prediction = agent.actor_model.forward(model_input)
+        log_probs = agent.action_distribution.log_prob(prediction, initial_action)
+        log_prob_diff = log_probs - initial_action_log_probs
+        ratio = jnp.exp(log_prob_diff)
+
+        # Andrychowicz (2021) did not find any benefit in minibatch normalization.
+        if self.normalize_advantage:
+            advantages = (advantages - advantages.mean()) / (advantages.std() + self.eps)
+
+        # Policy loss, with clipping
+        clipped_ratio = jnp.clip(ratio, 1 - self.clip_param, 1 + self.clip_param)
+        policy_objective = jnp.mean(
+            jnp.minimum(
+                ratio * advantages,
+                clipped_ratio * advantages,
+            )
+        )
+
+        # Value loss term
+        values = agent.critic_model.forward(model_input).squeeze(axis=-1)
+        # give axes
+        values = values.squeeze(axis=-1)  # (time, env)
+        value_mse = jax.lax.cond(
+            self.clip_value_loss,
+            lambda: 0.5
+            * self._clipped_value_loss(
+                target_values=initial_values,
+                values=values,
+                value_targets=value_targets,
+            ),
+            lambda: 0.5 * jnp.mean((value_targets - values) ** 2),
+        )
+        value_objective = self.config.value_loss_coef * value_mse
+
+        # Entropy bonus term.
+        entropies = agent.action_distribution.entropy(prediction, rng)
+        entropy_objective = self.entropy_coef * jnp.mean(entropies)
+
+        total_objective = policy_objective - value_objective + entropy_objective
+        total_loss = -total_objective
+
+        metrics_to_log: dict[str, Array] = {
+            "policy_objective": policy_objective,
+            "value_objective": value_objective,
+            "entropy_objective": entropy_objective,
+            "total_objective": total_objective,
+            "ratio_mean": jnp.mean(ratio),
+            "ratio_std": jnp.std(ratio),
+            "ratio_max": jnp.max(ratio),
+            "ratio_min": jnp.min(ratio),
+            "log_prob_diff_mean": jnp.mean(log_prob_diff),
+            "advantage_norm_mean": jnp.mean(advantages),
+            "action_mean": jnp.mean(initial_action),
+            "action_std": jnp.std(initial_action),
+            "action_max": jnp.max(initial_action),
+            "action_min": jnp.min(initial_action),
+            "prediction_mean": jnp.mean(prediction),
+            "prediction_std": jnp.std(prediction),
+            "log_prob_mean": jnp.mean(log_probs),
+            "log_prob_max": jnp.max(log_probs),
+            "log_prob_min": jnp.min(log_probs),
+            "values_std": jnp.std(values),
+            "values_mean": jnp.mean(values),
+            "obs_nans_ratio": xax.compute_nan_ratio(initial_action),
+            "action_nans_ratio": xax.compute_nan_ratio(initial_action),
+        }
```

Comment:
i think we should separate this out into a different class - have this function return the relevant tensors, then have a separate function